### PR TITLE
acl: binding rules evaluation

### PR DESCRIPTION
**Pull request by pkazmierczak**
`Binder` provides an interface for binding claims and ACL roles/policies of Nomad.

### CODE REVIEW

**Review by jrasell — CHANGES_REQUESTED** [review]
Looking great, thanks for this! I have made a couple of suggestions, the policy name list is the only blocker as the other items can be looked up as we see fit.

File: lib/auth/oidc/claims.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	}
+
+	list, err := extractListMappings(all, am.Config.ListClaimMappings)
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]interface{}{
+		"value": values,
+		"list":  list,
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply lib/auth/oidc/claims.go:63]: Good idea.

File: nomad/structs/acl.go [hunk 2]
```diff
@@ -829,6 +829,14 @@ func (a *ACLAuthMethodConfig) Copy() *ACLAuthMethodConfig {
 	return c
 }
 
+// ACLAuthClaims is the claim mapping of the OIDC auth method in a format that
+// can be used with go-bexpr. This structure is used during rule binding
+// evaluation.
+type ACLAuthClaims struct {
+	Value any
+	List  any
```
Comment (jrasell) [inline nomad/structs/acl.go:837]: I think we can make these map types, so we avoid asserting it each time to grab the data.
```suggestion
	Value map[string]string{}
	List  map[string][]string{}
```

Comment (pkazmierczak) [reply nomad/structs/acl.go:837]: Ah right I forgot these will always be maps 🤦‍♂️

**Pull request merged**